### PR TITLE
AMOEBA/HIPPO RMS force accuracy estimate

### PR DESCRIPTION
**Pull request by ndtrung81**
**Summary**

This PR attempts to give an estimate of the RMS force accuracy for the AMOEBA/HIPPO pair styles for a given grid and an alpha ewald value (both read in from the AM/HP .key files).

**Related Issue(s)**

N/A

**Author(s)**

Trung Nguyen (U Chicago)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

AMOEBA/HIPPO use multiple particle-mesh Ewald  solvers for multipole, polar and induced terms. As the first attempt, I chose the simplest approach using the monopoles (rpole[i][0]) and borrow how the RMS force accuracy is estimated in ewald.cpp. Estimates based on dipole could be possible, but will be more involving (like with pppm_dipole.cpp, minus  the P3M gf compute part). Importantly, I propose that the monopole-based accuracy is dominant relative to other terms.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/AMOEBA/pair_amoeba.cpp [hunk 1]
```diff
@@ -347,7 +347,7 @@ void PairAmoeba::compute(int eflag, int vflag)
     }
   }
 
-  first_flag_compute = 0;
+//  first_flag_compute = 0;
```
Comment (stanmoore1) [inline src/AMOEBA/pair_amoeba.cpp]: Remove

File: src/AMOEBA/pair_amoeba.cpp [hunk 2]
```diff
@@ -978,9 +984,9 @@ void PairAmoeba::init_style()
     for (int i = 0; i < nlocal; i++) pval[i] = 0.0;
   }
 
-  // output FF settings to screen and logfile
-
-  if (first_flag && (comm->me == 0)) print_settings();
+  // output FF settings to screen and logfile:
+  //   delay until rpole are available for RMS force accuracy estimate
+  //if (first_flag && (comm->me == 0)) print_settings();
```
Comment (stanmoore1) [inline src/AMOEBA/pair_amoeba.cpp]: Remove comment

File: src/AMOEBA/amoeba_multipole.cpp [hunk 3]
```diff
@@ -1000,3 +1001,61 @@ void PairAmoeba::damppole(double r, int rorder, double alphai, double alphak,
...
+  int nz_fft = m_kspace->nz;
+  double cutoff = mpolecut;
+
+  double lprx = rms(nx_fft,xprd,natoms,aeewald,q2);
+  double lpry = rms(ny_fft,yprd,natoms,aeewald,q2);
+  double lprz = rms(nz_fft,zprd_slab,natoms,aeewald,q2);
+  double lpr = sqrt(lprx*lprx + lpry*lpry + lprz*lprz) / sqrt(3.0);
+  double q2_over_sqrt = q2 / sqrt(natoms*cutoff*xprd*yprd*zprd_slab);
+  double spr = 2.0 *q2_over_sqrt * exp(-aeewald*aeewald*cutoff*cutoff);
+  double tpr = 0; //estimate_table_accuracy(q2_over_sqrt,spr);
```
Comment (stanmoore1) [inline src/AMOEBA/amoeba_multipole.cpp]: Remove commented out code

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
@ndtrung81 this looks good overall. Can you clean up the code a bit (see my comments)? Can you verify this is working correctly, e.g. using a highly refined calculation with a low error?

**Comment by ndtrung81** [comment]
@stanmoore1 thanks, I cleaned up the code as suggested. They were there for reviewing purposes, to explain how I made the changes.  Will need to do some tests regarding accuracy changing with mesh size. The mesh in the AMOEBA examples are now hard coded in the param files.

**Review by stanmoore1 — APPROVED** [review]

**Comment by ndtrung81** [comment]
I will do some more tests before assigning the PR to Axel.

**Comment by ndtrung81** [comment]
@sjplimp With the given metrics, the current mesh (60 48 48) gives a relative RMS force accuracy of about 1e-5. I experimented with different meshes and noticed that coarser meshes, (20 16 16) and (15 12 12) could give a decent accuracy of [1e-5;1e-4]. If this accuracy is acceptable, the reduction in the number of mesh points would help reduce the cost of the many kspace calculations remarkably.

**Review by sjplimp — APPROVED** [review]
@ndtrung81 Thanks for adding this - seems like a useful metric.

**Pull request merged**